backend/danswer/connectors/gmail/connector.py

```diff
@@ -0,0 +1,234 @@
+                GMAIL_DB_CREDENTIALS_DICT_SERVICE_ACCOUNT_KEY
+            ]
+            creds = get_gmail_creds_for_service_account(
+                service_account_key_json_str=service_account_key_json_str
+            )
+
+            # "Impersonate" a user if one is specified
+            delegated_user_email = cast(
+                str | None, credentials.get(DB_CREDENTIALS_DICT_DELEGATED_USER_KEY)
+            )
+            if delegated_user_email:
+                creds = creds.with_subject(delegated_user_email) if creds else None
+
+        if creds is None:
+            raise PermissionError(
+                "Unable to access Gmail - unknown credential structure."
+            )
+
+        self.creds = creds
+        return new_creds_dict
+
+    @staticmethod
+    def _strptime_to_datetime(_input: str) -> datetime:
+        _input = _input.split(", ")[-1]  # remove weekday if exist
+        _input = _input.rsplit(" ", 1)[0]  # remove timezone: GMT, (GMT), (UTC), etc.
+        if "+" not in _input and "-" not in _input:
+            return datetime.strptime(_input, "%d %b %Y %H:%M:%S")
+        return datetime.strptime(_input, "%d %b %Y %H:%M:%S %z")
+
+    def _get_email_body(self, payload) -> str:
+        parts = payload.get("parts", [])
+        email_body = ""
+        for part in parts:
+            mime_type = part.get("mimeType")
+            body = part.get("body")
+            if mime_type == "text/plain":
+                data = body.get("data", "")
+                text = urlsafe_b64decode(data).decode()
+                email_body += text
+        return email_body
+
+    def _email_to_document(self, full_email: Dict[str, Any]) -> Document:
+        email_id = full_email["id"]
+        payload = full_email["payload"]
+        headers = payload.get("headers")
+        if headers:
+            for header in headers:
+                name = header.get("name")
+                value = header.get("value")
+                if name.lower() == "from":
+                    _from = value
+                if name.lower() == "to":
+                    _to = value
+                if name.lower() == "subject":
+                    _subject = value
+                if name.lower() == "date":
+                    _date = value
+        email_data = f"{_from} {_to} {_subject} {_date}"
+        logger.debug(f"{email_data}")
+        email_body_text: str = self._get_email_body(payload)
+        email_updated_at = self._strptime_to_datetime(_date).astimezone(timezone.utc)
+        link = f"https://mail.google.com/mail/u/0/#inbox/{email_id}"
+        return Document(
+            id=email_id,
+            sections=[Section(link=link, text=email_data + email_body_text)],
+            source=DocumentSource.GMAIL,
+            title=_subject,
+            semantic_identifier=_subject,
+            doc_updated_at=email_updated_at,
+            metadata={} if email_data else {IGNORE_FOR_QA: True},
```

Comment:
Optional: maybe cool to have something like "flag: important" or "flag: starred" or whatever they have it as. Just food for thought, totally not required

Reply:
All labels are now included so metadata looks like:
```
{"labels": ["UNREAD", "IMPORTANT", "CATEGORY_UPDATES", "STARRED", "INBOX"]}
```
LMK what you think